crates/iceberg-rest-server/src/api/v1/warehouse.rs

```diff
@@ -75,6 +105,113 @@ pub trait WarehouseService<C: Catalog, A: AuthZHandler, S: SecretStore> {
             warehouse_id: warehouse_id.into_uuid(),
         })
     }
+
+    async fn update_warehouse(
+        request: UpdateWarehouseRequest,
+        context: ApiContext<State<A, C, S>>,
+        _headers: HeaderMap,
+    ) -> Result<UpdateWarehouseResponse> {
+        let warehouse_id = &WarehouseIdent::from(request.warehouse_id);
+        let mut transaction = C::Transaction::begin_write(context.v1_state.catalog.clone()).await?;
+        let old_warehouse_config =
+            C::get_warehouse(warehouse_id, context.v1_state.catalog.clone()).await?;
+        let (old_name, old_storage_profile, old_storage_credential_id) = {
+            (
+                old_warehouse_config.warehouse_name,
+                old_warehouse_config.storage_profile,
+                old_warehouse_config.storage_credential_id,
+            )
+        };
+
+        if let Some(new_name) = request.new_name {
+            if new_name != old_name {
+                C::update_warehouse_name(warehouse_id, &new_name, transaction.transaction())
+                    .await?;
+            }
+        }
+
+        match (request.new_storage_credential, request.new_storage_profile) {
+            (None, Some(_)) => {
+                let err: IcebergErrorResponse = ErrorModel::builder()
+                    .message("Cannot update 'storage_profile' without new creds.")
+                    .code(StatusCode::CONFLICT.into())
+                    .r#type("CannotUpdateWarehouse")
+                    .build()
+                    .into();
+
+                Err(err)
+            }
+            (None, None) => Ok(()),
+            (Some(new_storage_credential), Some(new_storage_profile)) => {
+                let mut new_storage_profile = new_storage_profile;
+                new_storage_profile
+                    .validate(Some(&new_storage_credential))
+                    .await?;
+
+                let can_be_updated = match (&new_storage_profile, old_storage_profile) {
+                    (StorageProfile::S3(new_profile), StorageProfile::S3(old_profile)) => {
+                        old_profile.can_be_updated_with(new_profile)
+                    }
+                };
+
+                if can_be_updated {
+                    C::update_warehouse_storage_profile(
+                        warehouse_id,
+                        new_storage_profile,
+                        transaction.transaction(),
+                    )
+                    .await?;
+                }
+
+                if let Some(old_cred) = old_storage_credential_id {
+                    S::delete_secret(&old_cred, context.v1_state.secrets.clone()).await?;
```

Comment:
This is an atomic transaction. We should delete the secret at the very end of the function - even after the Postgres Transaction has been committed. Otherwise we might end up with a corrupt state (i.e. old secret already deleted, warehouse commit failed, so that old warehouse commits now points to nonexistent secret).

Also use .ok() here, as the transaction can still succeed if deleting the secret fails.